docs/developer/adr/0002-remove-special-handling-of-optional-and-union.md

```diff
@@ -0,0 +1,92 @@
+Instead, an implicit `None` provider is added.
+This has a series of problems, which we exemplify for the case of `Optional`.
+
+1. Default values (which are currently ignored by Sciline) are overridden by the implicit `None` provider.
+   In other words, Sciline assumes that the default value of the optional input is `None`.
+2. Entire branches a pruned, which can hide bugs.
+   If the users added providers for the optional input, they will not be used if any of them has unintentionally unsatisfied dependencies.
+3. The special mechanism prevents the (in principle very valid) use of any providers that return an `Optional` or `Union` type.
+4. Optional inputs cannot be set to `None` *explicitly*.
+
+In summary, the special handling of `Optional` and `Union` is too implicit and causes more problems than it solves.
+There are a couple more aspects to consider.
+
+### Readability of user code
+
+Handling `Optional` explicitly would make user code more readable.
+Consider the following example:
+
+```python
+pipeline[MyParam] = 1.2
+```
+
+In the current implementation this gives no indication to the user that `MyParam` is not a required input.
+Furthermore, if the line is removed, the user may not realize that `MyParam` is available as an optional input.
+With the proposed change, the user can make this explicit:
+
+```python
+pipeline[Optional[MyParam]] = 1.2
+```
+
+Above it is clear that `MyParam` is optional, and it can be set to `None` explicitly:
+
+```python
+pipeline[Optional[MyParam]] = None
+```
+
+### Code complexity and maintainability
+
+The special handling of `Optional` and `Union` is a significant source of complexity in the code, requiring a significant amount of unit testing.
+
+### Conceptual clarity
+
+The current redesign of Sciline highlighted that the current implementation is conceptually flawed.
+It makes it tricky to represent the internals of `sciline.Pipeline` as a simple data dependency graph.
+The special handling of `Optional` and `Union` seems to require pervasive changes to the code, which is a sign that it is not a good fit for the design.
+
+### Counter arguments
+
+#### Multiple providers may depend on the same input, but not all optionally
+
+This seems like a special case that we have not seen in practice, is likely not worth the complexity of the current implementation.
+
+#### Using a provider returning a non-optional output to fulfill an optional input
+
+This is a very valid use case, but it would be made impossible if we stop associating a node `T` with an optional input `Optional[T]`.
+There are a couple of possible workarounds:
+
+- Add an explicit `Optional` provider that wraps (or depends on) the non-optional provider.
+- Modify the graph structure (which we plan to support in the redesign of Sciline) using something like `pipeline[Optional[MyParam]] = pipeline[MyParam]`.
```

Comment:
3rd option:
Since with the proposed approach, certain `T` basically always occur as `Optional[T]`, we could adapt the domain types to include `Optional` directly, e.g.
```python
MyParam = NewType('MyParam', Optional[float])
```
in the case above.

Reply:
Isn't that a sub case of the first option (just syntactic sugar)?